modalc-lib/modal-vector.rkt

```diff
@@ -0,0 +1,118 @@
+                                      (proj arg neg-party))
+                                    value]
+                                   [else (printf "no ref ~s\n" (list index value))
+                                         value]))
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc? (list index value))
+                                    (printf "set call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    (vector-set! vec index value)
+                                    value]
+                                   [else (printf "no set ~s\n" (list index value))
+                                         value])))))))
+
+(define-simple-macro (modal-vector/c* mode
+                                 {~and inner-ctc {~not _:keyword}} ...
+                                 {~seq kw:keyword ctc} ...)
+  (make-modal-vector/c* mode
+                   (list inner-ctc ...)
+                   (list (list 'kw ctc) ...)))
+
+(define (make-modal-vector/c* should-apply-ctc? inner-ctc kwds)
+  (define ref-ctc? should-apply-ctc?)
+  (define set-ctc? should-apply-ctc?)
+  (for ([kw kwds])
+    (cond [(equal? '#:ref (car kw))
+           (set! ref-ctc? (car (cdr kw)))]
+          [(equal? '#:set (car kw))
+           (set! set-ctc? (car (cdr kw)))]))
+  (define inner-ctc-proj
+    (map contract-late-neg-projection inner-ctc))
+  (make-contract
+   #:name `(modal-vector/c ,(map contract-name inner-ctc-proj))
+   #:late-neg-projection
+   (λ (blame)
+     (define inner-ctc-proj/blame (map (λ (p) (p (blame-swap blame))) inner-ctc-proj))
+     (λ (val neg-party)
+       (impersonate-vector val
+                           (λ (vec index value)
+                             (cond [(ref-ctc? (list index value))
+                                    (printf "ref call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    value]
+                                   [else (printf "no ref ~s\n" (list index value))
+                                         value]))
+                           (λ (vec index value)
+                             (cond [(set-ctc? (list index value))
+                                    (printf "set call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    (vector-set! vec index value)
```

Comment:
There's no need to do the `vector-set!` here, the idea of `impersonate-vector` here is that it will happen after this function gets called, using whatever this function returned.

Reply:
That's what I figured, but when I didn't have this there were errors with not updating the list as it should have been